src/display/KT-LCD3/eeprom.c

```diff
@@ -316,31 +316,69 @@ static void variables_to_array (uint8_t *ui8_array)
   ui8_array [62] = (p_configuration_variables->ui32_odometer_x10 >> 16) & 255;
 }
 
-static void eeprom_write_array (uint8_t *array, uint8_t ui8_len)
+static void eeprom_write_array (uint8_t *p_array, uint8_t ui8_len)
 {
   uint8_t ui8_i;
+  uint8_t ui8_data;
+  uint8_t *p_array_data;
+  uint8_t array_data_read_back [EEPROM_BYTES_STORED];
+  uint8_t ui8_data_written_correctly = 0;
 
   FLASH_SetProgrammingTime(FLASH_PROGRAMTIME_STANDARD);
   
   FLASH_Unlock (FLASH_MEMTYPE_DATA); // Unlock Data memory  
   while (FLASH_GetFlagStatus(FLASH_FLAG_DUL) == RESET) { } // Wait until Data EEPROM area unlocked flag is set
 
-  for (ui8_i = 0; ui8_i < ui8_len; ui8_i++)
+  // on next loop, we write the array to EEPROM and then read again and compared the values,
+  // if they are different, we keep writing until they are equal
+  do
   {
-    FLASH_ProgramByte (((uint32_t) EEPROM_BASE_ADDRESS) + ((uint32_t) ui8_i), *array++);
+    // write the full array
+    p_array_data = p_array;
+    for (ui8_i = 0; ui8_i < ui8_len; ui8_i++)
+    {
+      FLASH_ProgramByte (((uint32_t) EEPROM_BASE_ADDRESS) + ((uint32_t) ui8_i), *p_array_data++);
+    }
+
+    // read back the full array
+    variables_to_array (array_data_read_back);
+
+    // compare each byte of read array so see if the values were correctly written
+    ui8_data_written_correctly = 1;
+    p_array_data = p_array;
+    for (ui8_i = 0; ui8_i < ui8_len; ui8_i++)
+    {
+      ui8_data = *p_array_data++;
+      if (ui8_data != array_data_read_back[ui8_i])
+      {
+        ui8_data_written_correctly = 0;
+        break;
+      }
+    }
   }
+  while (ui8_data_written_correctly == 0); // loop until data is written correctly
```

Comment:
No infinite loop, code is working so far so good.

I never saw an incorrect written (I just saw that code trigger when I had a bug on it).

https://www.tutorialspoint.com/cprogramming/c_do_while_loop.htm
do {
   statement(s);
} while( condition );

Reply:
Nice. I know how a do-while loop works ;) Just want to make sure and get confirmation that ui8_data_written_correctly will become 1 otherwise it just hangs...